pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,416 @@
+
+def calc_cntrb_importance(df, action_type, threshold):
+    # sort rows in df based on number of contributions from greatest to least
+    df = df.sort_values(by=action_type, ascending=False)
+
+    # calculate the threshold amount of contributions 
+    thresh_cntrbs = df[action_type].sum() * threshold
+    
+    # initilize running sum of contributors who make up contributor importance
+    cntrb_importance = 0
+
+    # initialize running sum of contributions
+    running_sum = 0
+    
+    for _, row in df.iterrows():
+        running_sum += row[action_type] # update the running sum by the number of contributions a contributor has made
+        cntrb_importance+=1 # update contributor importance 
+        # if the running sum of contributions is greater than or equal to the threshold amount, break
+        if running_sum >= thresh_cntrbs:
+            break
+
+    return cntrb_importance
+
+def cntrb_importance_over_time(df, period_from, period_to, window_width, threshold):
+    # subset df such that the rows correspond to the window of time defined by period from and period to
+    mask = (df['created_at'] >= period_from) & (df['created_at'] <= period_to)
+    df_in_range = df.loc[mask]
+    
+    # count the number of contributions each contributor has made according each action type
+    df_count_cntrbs = df_in_range.groupby(['Action', 'cntrb_id'])['cntrb_id'].count().to_frame()
+    df_count_cntrbs = df_count_cntrbs.rename(columns={'cntrb_id': 'count'}).reset_index()
+     
+    # pivot df such that the column names correspond to the different action types and the values are the number of contributions of each contributor
+    df_count_cntrbs = df_count_cntrbs.pivot(columns='Action', values='count')
+
+    # initilize empty df with columns that correspond to the different action types to store contributor importance 
+    df_cntrb_importance = pd.DataFrame(columns=df_count_cntrbs.columns)
+    
+    # create a dictionary to store contributor importance for each action type
+    cis = {action_type: 0 for action_type in df_count_cntrbs.columns}
+    for action_type in df_count_cntrbs.columns:
+        # calculate contributor importance for each action time for the specified time window
+        ci = calc_cntrb_importance(df_count_cntrbs, action_type, threshold)
+        cis[action_type] = ci
+    
+    return cis
+
+def process_data(df, patterns, threshold, window_width, step_size, start_date, end_date):
+    # if the start_date and/or the end date is not specified set them to the beginning and most recent created_at date
+    if start_date is None:
+        start_date = df["created_at"].min()
+    if end_date is None:
+        end_date = df["created_at"].max()
+
+    if patterns:
+        # remove rows where Login column value contains the substring 'bot'
+        mask = df['login'].str.contains("|".join(patterns), na=False)
+        df = df[~mask]
+    
+    # convert integer threshold to a float decimal
+    threshold = threshold / 10
```

Comment:
use a python type conversion function instead of brute force

Reply:
I don't think I can use type conversion for division. The threshold is selected by the `Slider` object in Dash which only accepts integers as values eg) 10, 20, 30, etc. So I have to divide it by 10 to get the decimal representation of a percentage eg) 0.10, 0.20, 0.30, etc. I'll clarify my comment

Reply:
Sounds good, that makes a lot of sense with the additional context